data-prepper-plugins/saas-source-plugins/jira-source/src/main/java/org/opensearch/dataprepper/plugins/source/jira/JiraClient.java

```diff
@@ -21,25 +38,90 @@
-    public JiraClient() {
+    public JiraClient(JiraService service,
+                      JiraIterator jiraIterator,
+                      PluginExecutorServiceProvider executorServiceProvider,
+                      JiraSourceConfig sourceConfig) {
+        this.service = service;
+        this.jiraIterator = jiraIterator;
+        this.executorService = executorServiceProvider.get();
+        this.configuration = sourceConfig;
     }
 
-
     @Override
     public Iterator<ItemInfo> listItems() {
-        return null;
+        jiraIterator.initialize(lastPollTime);
+        return jiraIterator;
     }
 
     @Override
     public void setLastPollTime(Instant lastPollTime) {
-        log.info("Setting the lastPollTime: {}", lastPollTime);
+        log.trace("Setting the lastPollTime: {}", lastPollTime);
         this.lastPollTime = lastPollTime;
     }
 
+    @VisibleForTesting
+    void injectObjectMapper(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
     @Override
-    public void executePartition(SaasWorkerProgressState state, Buffer<Record<Event>> buffer, CrawlerSourceConfig configuration) {
-        log.info("Logic for executing the partitions");
+    public void executePartition(SaasWorkerProgressState state,
+                                 Buffer<Record<Event>> buffer,
+                                 CrawlerSourceConfig configuration) {
+        log.trace("Executing the partition: {} with {} ticket(s)",
+                state.getKeyAttributes(), state.getItemIds().size());
+        List<String> itemIds = state.getItemIds();
+        Map<String, Object> keyAttributes = state.getKeyAttributes();
+        String project = (String) keyAttributes.get(PROJECT);
+        Instant eventTime = state.getExportStartTime();
+        List<ItemInfo> itemInfos = new ArrayList<>();
+        for (String itemId : itemIds) {
+            if (itemId == null) {
+                continue;
+            }
+            ItemInfo itemInfo = JiraItemInfo.builder()
+                    .withItemId(itemId)
+                    .withId(itemId)
+                    .withProject(project)
+                    .withEventTime(eventTime)
+                    .withMetadata(keyAttributes).build();
+            itemInfos.add(itemInfo);
+        }
+
+        List<Record<Event>> recordsToWrite = itemInfos
+                .parallelStream()
+                .map(t -> (Supplier<String>) (() -> service.getIssue(t.getId())))
+                .map(supplier -> supplyAsync(supplier, this.executorService))
+                .map(CompletableFuture::join)
+                .map(ticketJson -> {
+                    try {
+                        return objectMapper.readValue(ticketJson, new TypeReference<>() {
+                        });
+                    } catch (JsonProcessingException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                .map(t -> (Event) JacksonEvent.builder()
```

Comment:
We are moving toward EventFactory. Please fix this in a follow-on PR.

Reply:
Ok 👍